### PR TITLE
Resend blocks when modified while sending to client

### PR DESCRIPTION
**Pull request by Jeija**
From what I understand, this can happen when the server sends node updates to the client:
- More than 4 nodes change at the same time (server.cpp:888)
- This causes the server to only send nodes that are 5 nodes or less away from the player (server.cpp:911), for any other nodes the corresponding position in `m_blocks_sent` in `RemoteClient` will be removed to mark those blocks for resending.
- `Server::SendBlocks` calls `RemoteClient::GetNextBlocks` which returns among others the block that was just removed from `m_blocks_sent`
- Here is the problem: While the block is transferred to the client, _the block on the server is modified again_. The corresponding position in `m_blocks_sent` is marked for resending again.
- Before the server can send the new blocks, the client answers with `TOSERVER_GOTBLOCKS`, which causes the block in `m_blocks_sent` to be unmarked (=marked as sent = inserted into `m_blocks_sent`):

``` C++
void RemoteClient::GotBlock(v3s16 p)
{
    /** some other stuff here **/
    m_blocks_sent.insert(p);
}
```
- Result: The server thinks the client's copy of the block is up to date when in fact, the block has been modified in the meantime.

You might think this issue is completely hypothetical due to the short time between sending the block and receiving the answer from the client, but it causes major issues when using mods that modify many nodes at the same time, such as mesecons:
#### These are not mesecons bugs, but caused by minetest itself:

You can easily reproduce this issue by building a small mesecons circuit, most annoyingly using pistons. You will get results like these (screenshots were taken in singleplayer):
screenshot_20160215_164205
screenshot_20160215_170452
And it is not just a matter of waiting a few seconds for the server to resend the block, you have to manually cause a block update in order to see what is really going on.
#### Solution

My patch makes the server remember which blocks were modified, so it can ignore the `TOSERVER_GOTBLOCKS` answers from the client for those blocks that changed since last sending an update.

### CODE REVIEW

**Comment by Megaf** [comment]
Sounds great! I will give it a try and report back later.
Thanks!

File: src/clientiface.cpp [hunk 1]
```diff
@@ -370,17 +370,24 @@ void RemoteClient::GetNextBlocks (
...
+		}
+
+		m_blocks_sent.insert(p);
 	}
-	m_blocks_sent.insert(p);
 }
 
 void RemoteClient::SentBlock(v3s16 p)
 {
+	if(m_blocks_modified.find(p) != m_blocks_modified.end())
```
Comment (est31) [inline src/clientiface.cpp]: Space between `if` and (

File: src/clientiface.cpp [hunk 2]
```diff
@@ -370,17 +370,24 @@ void RemoteClient::GetNextBlocks (
 
 void RemoteClient::GotBlock(v3s16 p)
 {
-	if(m_blocks_sending.find(p) != m_blocks_sending.end())
-		m_blocks_sending.erase(p);
-	else
+	if (m_blocks_modified.find(p) == m_blocks_modified.end())
 	{
-		m_excess_gotblocks++;
+		if(m_blocks_sending.find(p) != m_blocks_sending.end())
```
Comment (est31) [inline src/clientiface.cpp]: Space between `if` and (

File: src/clientiface.cpp [hunk 3]
```diff
@@ -370,17 +370,24 @@ void RemoteClient::GetNextBlocks (
...
+			m_blocks_sending.erase(p);
+		else
```
Comment (est31) [inline src/clientiface.cpp]: put the brace on the same line as the else, separed by one space

File: src/clientiface.cpp [hunk 4]
```diff
@@ -370,17 +370,24 @@ void RemoteClient::GetNextBlocks (
 
 void RemoteClient::GotBlock(v3s16 p)
 {
-	if(m_blocks_sending.find(p) != m_blocks_sending.end())
-		m_blocks_sending.erase(p);
-	else
+	if (m_blocks_modified.find(p) == m_blocks_modified.end())
```
Comment (est31) [inline src/clientiface.cpp]: brace on same line as ).

File: src/clientiface.h [hunk 5]
```diff
@@ -396,6 +396,16 @@ class RemoteClient
 	std::map<v3s16, float> m_blocks_sending;
 
 	/*
+		Blocks that have been modified since last sending them.
+		These blocks will not be marked as sent, even if the
+		client reports it has received them to account for blocks
+		that are being modified while on the line.
+
+		Key is position, value is dummy
```
Comment (est31) [inline src/clientiface.h]: No need to replicate the `value is dummy` for `m_blocks_sent`. It is a mistake left over from commit 6a1670dbc31cc0e44178bbd9ad34ff0d5981a060 (the commit replaced map with set but didnt update the comment).

**Comment by est31** [comment]
See comments above. Otherwise looks good.

**Comment by Jeija** [comment]
Thank you @est31, I have updated my first commit according to your suggestions.
While doing that though, I found many more the style issues in `clientiface.cpp` (actually there are more `if(` than `if (`). So I added a second commit to this PR that would fix all those issues.

So, feel free to merge only the first commit, it contains the style fixes to my code you suggested, the second just contains some more for the rest of the code.

**Comment by est31** [comment]
Its a good idea to have it merged as two commits.

**Comment by kilbith** [comment]
@est31 0.4.14 milestone ?

**Comment by orwell96** [comment]
We definitely need this.
I lately recognized a bug with placing locked chests or other types of nodes that are setting up their inventory in after_place_node callback which appears since [my PR](https://github.com/minetest/minetest/commit/e17fbb31d6db3e6d2d9b68f5793ad8b2146f7bcc) regarding m_nothing_to_send_pause_timer.
1. client sends: placing chest
2. server queues mapblock for resending immediately
3. server calls after_place node
4. client answers with gotblocks and server thinks the client already knows about the inventory being created.
Result: The client does not know about the inventory and shows an empty formspec.
Ironically, i initially set m_nothing_to_send_pause_timer to 0.1 instead of 0.

**Comment by est31** [comment]
Okay, nore has +1ed the PR, see IRC log: http://irc.minetest.ru/minetest-dev/2016-03-12#i_4555202

Merged as 089f9bbe817debc5c30f1d845170909f638c9642
only the first commit. @Jeija I couldn't apply the second commit patch, can you rebase it and add it in a separate PR?

**Pull request closed without merging**